functions/Get-DbaRegisteredServer.ps1

```diff
@@ -0,0 +1,159 @@
+
+			Gets a list of servers from the CMS on sqlserver2014a, using Windows Credentials.
+
+		.EXAMPLE
+			Get-DbaRegisteredServer -SqlInstance sqlserver2014a -SqlCredential $credential | Select-Object -Unique -ExpandProperty ServerName
+
+			Returns only the server names from the CMS on sqlserver2014a, using SQL Authentication
+
+		.EXAMPLE
+			Get-DbaRegisteredServer -SqlInstance sqlserver2014a -Group HR, Accounting
+
+			Gets a list of servers in the HR and Accounting groups from the CMS on sqlserver2014a.
+
+		.EXAMPLE
+			Get-DbaRegisteredServer -SqlInstance sqlserver2014a -Group HR\Development
+
+			Returns a list of servers in the HR and sub-group Development from the CMS on sqlserver2014a
+	#>
+	[CmdletBinding(DefaultParameterSetName = "Default")]
+	param (
+		[parameter(Mandatory = $true, ValueFromPipeline = $true)]
+		[Alias("ServerInstance", "SqlServer")]
+		[DbaInstanceParameter[]]$SqlInstance,
+		[PSCredential]$SqlCredential,
+		[Alias("Groups")]
+		[object[]]$Group,
+		[object[]]$ExcludeGroup,
+		[switch]$ExcludeCmsServer,
+		[switch]$Silent
+	)
+	begin {
+		function Find-CmsGroup {
+			[OutputType([object[]])]
+			[cmdletbinding()]
+			param(
+				$CmsGrp,
+				$Base = $null,
+				$Stopat
+			)
+			$results = @()
+			foreach ($el in $CmsGrp) {
+				if ($null -eq $Base -or [string]::IsNullOrWhiteSpace($Base) ) {
+					$partial = $el.name
+				}
+				else {
+					$partial = "$Base\$($el.name)"
+				}
+				if ($partial -eq $Stopat) {
+					return $el
+				}
+				else {
+					foreach ($elg in $el.ServerGroups) {
+						$results += Find-CmsGroup -CmsGrp $elg -Base $partial -Stopat $Stopat
+					}
+				}
+			}
+			return $results
+		}
+	}
+
+	process {
+		if (Test-FunctionInterrupt) { 
+			return 
+		}
+
+		$servers = @()
+		foreach ($instance in $SqlInstance) {
+			try {
+				$cmsStore = Get-DbaRegisteredServersStore -SqlInstance $instance -SqlCredential $SqlCredential -Silent:$Silent
+			}
+			catch {
+				Stop-Function -Message "Cannot access Central Management Server '$instance'." -ErrorRecord $_ -Continue
+				return
+			}
+
+			if (Test-Bound -ParameterName ExcludeGroup) {
+				$Group = ($cmsStore.DatabaseEngineServerGroup.ServerGroups | Where-Object Name -notin $ExcludeGroup).Name
+			}
+
+			if ($Group) {
+				foreach ($currentGroup in $Group) {
+					$cms = Find-CmsGroup -CmsGrp $cmsStore.DatabaseEngineServerGroup.ServerGroups -Stopat $currentGroup
+					if ($null -eq $cms) {
+						Write-Message -Level Output -Message "No groups found matching that name on instance '$instance'."
+						continue
+					}
+					$servers += ($cms.GetDescendantRegisteredServers())
+				}
+			}
+			else {
+				$cms = $cmsStore.DatabaseEngineServerGroup
+				$servers += ($cms.GetDescendantRegisteredServers())
+			}
+
+			if ($ExcludeCmsServer) {
+				$servers = ($servers | Where-Object { $_.ServerName -ne $instance})
+			}
+		}
+
+		foreach ($server in $servers) {
+			try {
+				$lookup = Resolve-DbaNetworkName $server.ServerName -Turbo -Silent
```

Comment:
DNS resolution adds a lot of execution time, please keep in the `if ($NetBiosName -or $IpAddress) {` brackets.

Reply:
this still stands true - it needs to be an explicit request and not resolve by default. I have 21 servers in my CMS and the previous code that resolved by default went from milliseconds to like 10 seconds to run.

Reply:
That seems awfully slow. I ran a test on my group (of 105 instances), and doing the whole thing only added about 2.3 seconds to do the resolve (doing a subset of a few instances around the same size as your set added about the same amount for me...)

```
C:\DBA\DDL> Measure-Command{ $servers = Get-DbaRegisteredServer -SqlInstance sqldbatools.my.corp }


Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 5
Milliseconds      : 897
Ticks             : 58978397
TotalDays         : 6.82620335648148E-05
TotalHours        : 0.00163828880555556
TotalMinutes      : 0.0982973283333333
TotalSeconds      : 5.8978397
TotalMilliseconds : 5897.8397



C:\DBA\DDL> Measure-Command{ $servers = Get-DbaRegisteredServer -SqlInstance sqldbatools.my.corp -ResolveDns }


Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 8
Milliseconds      : 154
Ticks             : 81547386
TotalDays         : 9.43835486111111E-05
TotalHours        : 0.00226520516666667
TotalMinutes      : 0.13591231
TotalSeconds      : 8.1547386
TotalMilliseconds : 8154.7386



C:\DBA\DDL> $servers.Count
105
C:\DBA\DDL>
```

Would having it as a switch `-ResolveDns` (or `-ExcludeDns` that does the resolve by default) that adds the two properties I was pulling from `Resolve-DbaNetworkName` (IPAddress and ComputerName) be a good compromise?